drivers/dose/dose.c

```diff
@@ -0,0 +1,566 @@
+        case DOSE_STATE_SEND + DOSE_SIGNAL_XTIMER:
+            new_state = DOSE_STATE_BLOCKED;
+            break;
+        case DOSE_STATE_SEND + DOSE_SIGNAL_END:
+            new_state = DOSE_STATE_BLOCKED;
+            break;
+    }
+
+    assert(new_state > DOSE_STATE_UNDEF && new_state < DOSE_STATE_INVALID);
+
+    /* Call state specific function */
+    while (1) {
+        uint8_t next_state = new_state;
+        switch (next_state) {
+            case DOSE_STATE_BLOCKED: next_state = state_blocked(ctx, old_state); break;
+            case DOSE_STATE_RECV: next_state = state_recv(ctx, old_state); break;
+            case DOSE_STATE_SEND: next_state = state_send(ctx, old_state); break;
+        }
+
+        if (next_state == new_state) {
+            /* No state change occurred within the state's function */
+            break;
+        }
+        else {
+            old_state = new_state;
+            new_state = next_state;
+        }
+    }
+
+    assert(new_state > DOSE_STATE_UNDEF && new_state < DOSE_STATE_INVALID);
+
+    /* Indicate state change by unlocking state mutex */
+    ctx->state = new_state;
+    mutex_unlock(&ctx->state_mtx);
+    irq_restore(irq_state);
+    return ctx->state;
+}
+
+static void _isr_uart(void *arg, uint8_t c)
+{
+    dose_t *dev = (dose_t *) arg;
+
+    dev->uart_octet = c;
+    state(dev, DOSE_SIGNAL_UART);
+}
+
+static void _isr_gpio(void *arg)
+{
+    dose_t *dev = (dose_t *) arg;
+
+    state(dev, DOSE_SIGNAL_GPIO);
+}
+
+static void _isr_xtimer(void *arg)
+{
+    dose_t *dev = (dose_t *) arg;
+
+    state(dev, DOSE_SIGNAL_XTIMER);
+}
+
+static void clear_recv_buf(dose_t *ctx)
+{
+    int irq_state = irq_disable();
+
+    ctx->recv_buf_ptr = 0;
+    CLRBIT(ctx->flags, DOSE_FLAG_RECV_BUF_DIRTY);
+    CLRBIT(ctx->flags, DOSE_FLAG_END_RECEIVED);
+    CLRBIT(ctx->flags, DOSE_FLAG_ESC_RECEIVED);
+    irq_restore(irq_state);
+}
+
+static void _isr(netdev_t *netdev)
+{
+    dose_t *ctx = (dose_t *) netdev;
+    int irq_state, dirty, end;
+
+    /* Get current flags atomically */
+    irq_state = irq_disable();
+    dirty = (ctx->flags & DOSE_FLAG_RECV_BUF_DIRTY);
+    end = (ctx->flags & DOSE_FLAG_END_RECEIVED);
+    irq_restore(irq_state);
+
+    /* If the receive buffer does not contain any data just abort ... */
+    if (!dirty) {
+        DEBUG("dose _isr(): no frame -> drop\n");
+        return;
+    }
+
+    /* If we haven't received a valid END octet just drop the incomplete frame. */
+    if (!end) {
+        DEBUG("dose _isr(): incomplete frame -> drop\n");
+        clear_recv_buf(ctx);
+        return;
+    }
+
+    /* The set dirty flag prevents recv_buf or recv_buf_ptr from being
+     * touched in ISR context. Thus, it is safe to work with them without
+     * IRQs being disabled or mutexes being locked. */
+
+    /* Check the dst mac addr if the iface is not in promiscuous mode */
+    if (!(ctx->opts & DOSE_OPT_PROMISCUOUS)) {
+        ethernet_hdr_t *hdr = (ethernet_hdr_t *) ctx->recv_buf;
```

Comment:
We haven't checked if enough bytes have been received. We need at least `sizeof(ethernet_hdr_t) + DOSE_FRAME_CRC_LEN` bytes.